src/box/memtx_sort_data.c

```diff
@@ -0,0 +1,720 @@
+	writer->curr_entry =
+		mh_memtx_sort_data_writer_entries_node(writer->entries, i);
+	return 0;
+}
+
+int
+memtx_sort_data_writer_put(struct memtx_sort_data_writer *writer,
+			   void *data, size_t size, size_t count)
+{
+	assert(writer->curr_entry != NULL);
+	if (fwrite(data, size, count, writer->fp) != count) {
+		diag_set(SystemError, "%s: failed to write the sort data",
+			 writer->filename);
+		return -1;
+	}
+	writer->curr_entry->psize += size * count;
+	writer->curr_entry->len += count;
+	return 0;
+}
+
+int
+memtx_sort_data_writer_commit(struct memtx_sort_data_writer *writer)
+{
+	/* Update the dummy sort data entry in the file header. */
+	assert(writer->curr_entry != NULL);
+	struct memtx_sort_data_writer_entry *entry = writer->curr_entry;
+	if (writer_seek(writer, entry->header_entry_offset, SEEK_SET) != 0 ||
+	    writer_print(writer, ENTRY_FMT, entry->key.space_id,
+			 entry->key.index_id, entry->offset,
+			 entry->psize, entry->len) != 0)
+		return -1;
+	writer->curr_entry = NULL;
+	return 0;
+}
+
+int
+memtx_sort_data_writer_begin_pk(struct memtx_sort_data_writer *writer,
+				uint32_t space_id)
+{
+	return memtx_sort_data_writer_begin(writer, space_id, 0);
+}
+
+int
+memtx_sort_data_writer_put_pk_tuple(struct memtx_sort_data_writer *writer,
+				    struct tuple *tuple)
+{
+	return memtx_sort_data_writer_put(writer, &tuple, sizeof(tuple), 1);
+}
+
+int
+memtx_sort_data_writer_commit_pk(struct memtx_sort_data_writer *writer)
+{
+	return memtx_sort_data_writer_commit(writer);
+}
+
+/* }}} */
+
+/* {{{ Reader *****************************************************************/
+
+/**
+ * Read the expected next contents in the file header. Returns a pointer to the
+ * string right after the expected one (the buffer is statically allocated).
+ *
+ * WARNING: the max size of the header line is limited.
+ */
+static char *
+memtx_sort_data_header_expect(struct memtx_sort_data_reader *reader,
+			      const char *expect)
+{
+	char *line = tt_static_buf();
+	size_t expect_len = strlen(expect);
+	assert(expect_len < TT_STATIC_BUF_LEN);
+	if (fgets(line, TT_STATIC_BUF_LEN, reader->fp) == NULL ||
+	    memcmp(line, expect, expect_len) != 0) {
+		say_error("%s: file header read failed", reader->filename);
+		return NULL;
+	}
+	return line + expect_len;
+}
+
+struct memtx_sort_data_reader *
+memtx_sort_data_reader_new(const char *snap_filename,
+			   const struct vclock *vclock,
+			   const struct tt_uuid *instance_uuid)
+{
+	const char *filename = memtx_sort_data_filename(snap_filename);
+	struct memtx_sort_data_reader *reader = xmalloc(sizeof(*reader));
+	snprintf(reader->filename, sizeof(reader->filename), "%s", filename);
+	reader->entries = mh_memtx_sort_data_entries_new();
+	reader->curr_entry = NULL;
+	reader->old2new = NULL;
+
+	/* Open the .sortdata file. */
+	reader->fp = fopen(reader->filename, "rb");
+	if (reader->fp == NULL) {
+		say_error("%s: file open failed", reader->filename);
+		goto fail_free;
```

Comment:
It causes crash (https://gist.github.com/drewdzzz/7994f0a4512f9abc3dd8da0d5aae5887)

The buffer is not initialized, and you are trying to `free` it. Is it covered with a test case?